jina/types/arrays/neural_ops.py

```diff
@@ -52,29 +53,117 @@ def match(
+                                all values will be rescaled into range `[a, b]`.
+        :param metric_name: if provided, then match result will be marked with this string.
+        """
+
+        x_mat = np.stack(self.get_attributes('embedding'))
+        y_mat = np.stack(darray.get_attributes('embedding'))
+
+        dists = cdist(x_mat, y_mat, metric_name)
         dist, idx = top_k(dists, min(limit, len(darray)), descending=False)
         if normalization is not None:
             if isinstance(normalization, (tuple, list)):
                 dist = minmax_normalize(dist, normalization)
 
+        for _q, _ids, _dists in zip(self, idx, dist):
+            _q.matches.clear()
+            for _id, _dist in zip(_ids, _dists):
+                # Note, when match self with other, or both of them share the same Document
+                # we might have recursive matches .
+                # checkout https://github.com/jina-ai/jina/issues/3034
+                d = darray[int(_id)]
+                if d.id in self:
+                    d = Document(d, copy=True)
+                    d.pop('matches')
+                _q.matches.append(d, scores={metric_name: _dist}, copy=False)
+
+    def _match_online(
+        self, darray, cdist, limit, normalization, metric_name, batch_size
+    ):
+        """
+        Computes the matches between self and `darray` loading `darray` into main memory in chunks of size `batch_size`.
+
+        :param darray: the other DocumentArray or DocumentArrayMemmap to match against
+        :param cdist: the distance metric
+        :param limit: the maximum number of matches, when not given
+                      all Documents in `another` are considered as matches
+        :param normalization: a tuple [a, b] to be used with min-max normalization,
+                                the min distance will be rescaled to `a`, the max distance will be rescaled to `b`
+                                all values will be rescaled into range `[a, b]`.
+        :param metric_name: if provided, then match result will be marked with this string.
+        :param batch_size: length of the chunks loaded into memory from darray.
+        """
+
+        x_mat = np.stack(self.get_attributes('embedding'))
+        n_x = x_mat.shape[0]
+
+        def batch_generator(y_darray: 'DocumentArrayMemmap', n_batch: int):
+            for i in range(0, len(y_darray), n_batch):
+                y_mat = np.vstack(
+                    [y_darray[j].embedding for j in range(i, i + n_batch)]
+                )
+
+                yield y_mat, i
+
+        y_batch_generator = batch_generator(darray, batch_size)
+        top_dists = np.inf * np.ones((n_x, limit))
+        top_inds = np.zeros((n_x, limit), dtype=int)
+
+        for ybatch, ybatch_start_pos in y_batch_generator:
+            distances = cdist(x_mat, ybatch, metric_name)
+            dists, inds = top_k(distances, limit, descending=False)
+
+            if normalization is not None:
```

Comment:
put in one line
```suggestion
            if normalization is not None and isinstance(normalization, (tuple, list)):
```

Reply:
Or even `isinstance` already contains the first condition

Reply:
done